accepted/wasm-gc.md

```diff
@@ -0,0 +1,1362 @@
+wrappers around an internal reference so that we don't rely on `dyn Any` which
+requires manipulation from Rust code.
+
+## Defining the Pluggable GC Interface
+
+Plugging in different GC backends will require implementing two traits:
+
+1. `wasmtime_environ::GcCompiler`
+
+2. `wasmtime_runtime::GcRuntime`
+
+These traits let the collector hook into the compiler and runtime respectively.
+
+Note that these traits are not exposed from the main, user-facing `wasmtime`
+crate. We do not intend for these traits to have the same stability guarantees
+that the public `wasmtime` crate has. These traits are purely for internal
+factoring and running experiments with new collectors. The `wasmtime` crate will
+expose a small handful of built-in collectors to choose from; it won't let you
+bring your own collector. This is similar to the approach taken by the
+`wasmtime_runtime::InstanceAllocator` trait.
+
+Note that the interface will fix some design decisions for GC implementations:
+
+1. GC roots on the Wasm stack will always be precisely traced in a
+   stop-the-world phase. We won't do conservative stack scanning, nor will we
+   reference count Wasm-stack roots, nor will we incrementally trace stack
+   roots.
+
+2. GC roots held from Rust (either on the stack or some heap-allocated
+   structure) will always be reference counted, as described above in the public
+   API section.
+
+In fact, the GC implementation will not interact with roots directly at all, it
+will just be handed the root set at the start of a collection.
+
+### The `wasmtime_environ::GcCompiler` Trait
+
+The `GcCompiler` trait is used when compiling Wasm code. We will add a
+`gc_compiler` method to `wasmtime_environ::FuncEnvironment` to get the
+`GcCompiler` for the current function compilation:
+
+```rust
+trait FuncEnvironment {
+    // <existing methods>
+
+    type Gc: GcCompiler;
+    fn gc_compiler(&mut self) -> &mut Self::Gc;
+}
+```
+
+Fundamentally, the `GcCompiler` trait allows the collector to do three things:
+
+1. Define the layout for GC objects.
+
+2. Emit read and write barriers for struct field and array element accesses in
+   Wasm code, if necessary.
+
+3. Emit code sequences to get the type id of an object. We will use this when
+   implementing dynamic checks for downcasting references.
+
+I've sketched the trait and related types below, but I fully expect this to
+change during implementation when faced with reality. I do not expect those
+three bits of functionality to fundamentally change, however, even if we
+completely change or remove `GcStructLayout`, for example.
+
+```rust
+/// The layout of a GC object.
+pub enum GcObjectLayout {
+    Struct(GcStructLayout),
+    Array(GcArrayLayout),
+}
+
+/// The layout for a GC-managed struct type.
+pub struct GcStructLayout {
+    /// The size of this struct.
+    pub size: u32,
+    /// The alignment of this struct.
+    pub align: u32,
```

Comment:
In V8, object alignment is a global constant (and equals the pointer size), rather than being stored per object layout. That makes some things simpler, but sometimes we do wish we had the infrastructure to have per-object alignment requirements, so it'll be interesting to see whether you find that having this flexibility is worthwhile.